internal/proxy/proxy.go

```diff
@@ -54,44 +55,110 @@ type Config struct {
 // Client represents the state of the current instantiation of the proxy.
 type Client struct {
 	cmd    *cobra.Command
-	dialer *cloudsqlconn.Dialer
+	dialer DBDialer
 
 	// mnts is a list of all mounted sockets for this client
 	mnts []*socketMount
 }
 
-// NewClient completes the initial setup required to get the proxy to a "steady" state.
-func NewClient(ctx context.Context, cmd *cobra.Command, conf *Config) (*Client, error) {
-	d, err := cloudsqlconn.NewDialer(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("error initializing dialer: %v", err)
+type portConfig struct {
+	global    int
+	postgres  int
+	mysql     int
+	sqlserver int
+}
+
+func newPortConfig(global int) *portConfig {
+	return &portConfig{
+		global:    global,
+		postgres:  5432,
+		mysql:     3306,
+		sqlserver: 1433,
+	}
+}
+
+// nextPort returns the next port based on the initial global value.
+func (c *portConfig) nextPort() int {
+	p := c.global
+	c.global++
+	return p
+}
+
+func (c *portConfig) nextDBPort(version string) int {
+	switch {
+	case strings.HasPrefix(version, "MYSQL"):
+		p := c.mysql
+		c.mysql++
+		return p
+	case strings.HasPrefix(version, "POSTGRES"):
+		p := c.postgres
+		c.postgres++
+		return p
+	case strings.HasPrefix(version, "SQLSERVER"):
+		p := c.sqlserver
+		c.sqlserver++
+		return p
+	default:
+		// Unexpected engine version, use global port setting instead.
+		return c.nextPort()
 	}
-	c := &Client{cmd: cmd, dialer: d}
+}
 
-	port := conf.Port
-	var inc int
+// DBDialer is a dialer for a Cloud SQL instance.
+type DBDialer interface {
```

Comment:
nit: Should it be DbDialer?

Reply:
DBD is a bit hard to read. We could just call it a Dialer.

Reply:
SGTM